File: socnetscraper/store.py
from __future__ import annotations

import csv
import json
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Any

from socnetscraper.config import POSTS_CSV, POSTS_JSONL, RUNS_DIR

CSV_FIELDS = [
    "id",
    "code",
    "url",
    "username",
    "text",
    "published_at",
    "like_count",
    "reply_count",
    "repost_count",
    "query",
    "source",
    "scraped_at",
]


def _key(post: dict[str, Any]) -> str:
    return str(post.get("id") or post.get("code") or post.get("url") or "")


def parse_datetime(value: Any) -> datetime | None:
    if value in (None, ""):
        return None
    if isinstance(value, datetime):
        dt = value
    else:
        text = str(value).strip().replace("Z", "+00:00")
        try:
            dt = datetime.fromisoformat(text)
        except ValueError:
            return None
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt.astimezone(timezone.utc)


def post_datetime(post: dict[str, Any]) -> datetime | None:
    return parse_datetime(post.get("published_at"))


def within_lookback(post: dict[str, Any], hours: int) -> bool:
    dt = post_datetime(post)
    if dt is None:
        return False
    return dt >= datetime.now(timezone.utc) - timedelta(hours=max(1, hours))


def lookback_posts(posts: list[dict[str, Any]], hours: int) -> list[dict[str, Any]]:
    return [post for post in posts if within_lookback(post, hours)]


def load_existing() -> dict[str, dict[str, Any]]:
    posts: dict[str, dict[str, Any]] = {}
    if not POSTS_JSONL.exists():
        return posts
    with POSTS_JSONL.open(encoding="utf-8") as handle:
        for line in handle:
            line = line.strip()
            if not line:
                continue
            item = json.loads(line)
            key = _key(item)
            if key:
                posts[key] = item
    return posts


def merge_posts(
    existing: dict[str, dict[str, Any]], incoming: list[dict[str, Any]]
) -> tuple[list[dict[str, Any]], int]:
    new_count = 0
    for post in incoming:
        key = _key(post)
        if not key:
            continue
        previous = existing.get(key) or {}
        if key not in existing:
            new_count += 1
        merged = {**previous, **post}
        if not (post.get("text") or "").strip() and previous.get("text"):
            merged["text"] = previous["text"]
        if not (post.get("image_urls") or post.get("images")):
            if previous.get("image_urls"):
                merged["image_urls"] = previous["image_urls"]
            if previous.get("images"):
                merged["images"] = previous["images"]
        if not post.get("avatar") and previous.get("avatar"):
            merged["avatar"] = previous["avatar"]
        if not (post.get("parent") or {}).get("text") and (previous.get("parent") or {}).get("text"):
            merged["parent"] = previous["parent"]
            merged["is_reply"] = True
        existing[key] = merged
    ordered = sorted(
        existing.values(),
        key=lambda item: str(item.get("published_at") or item.get("scraped_at") or ""),
        reverse=True,
    )
    return ordered, new_count


def save_posts(posts: list[dict[str, Any]]) -> None:
    POSTS_JSONL.parent.mkdir(parents=True, exist_ok=True)
    with POSTS_JSONL.open("w", encoding="utf-8") as handle:
        for post in posts:
            handle.write(json.dumps(post, ensure_ascii=False) + "\n")
    with POSTS_CSV.open("w", encoding="utf-8", newline="") as handle:
        writer = csv.DictWriter(handle, fieldnames=CSV_FIELDS, extrasaction="ignore")
        writer.writeheader()
        for post in posts:
            writer.writerow({field: post.get(field, "") for field in CSV_FIELDS})


def save_run(summary: dict[str, Any]) -> Path:
    stamp = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H%M%SZ")
    path = RUNS_DIR / f"{stamp}.json"
    path.write_text(json.dumps(summary, ensure_ascii=False, indent=2), encoding="utf-8")
    return path

File: plugins/mcp/src/scraper/store.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { paths } from "./config.js";
import type { Post } from "./parse.js";

function key(post: Post): string {
  return String(post.id || post.code || post.url || "");
}

export function loadExisting(root: string): Record<string, Post> {
  const file = paths(root).postsJsonl;
  const posts: Record<string, Post> = {};
  if (!existsSync(file)) {
    return posts;
  }
  for (const line of readFileSync(file, "utf8").split(/\r?\n/)) {
    if (!line.trim()) {
      continue;
    }
    const item = JSON.parse(line) as Post;
    const id = key(item);
    if (id) {
      posts[id] = item;
    }
  }
  return posts;
}

export function mergePosts(existing: Record<string, Post>, incoming: Post[]): [Post[], number] {
  let newCount = 0;
  for (const post of incoming) {
    const id = key(post);
    if (!id) {
      continue;
    }
    const previous = existing[id] || {};
    if (!existing[id]) {
      newCount += 1;
    }
    const merged: Post = { ...previous, ...post };
    if (!String(post.text || "").trim() && previous.text) {
      merged.text = previous.text;
    }
    const prevParent = previous.parent as Post | undefined;
    const nextParent = post.parent as Post | undefined;
    if (!nextParent?.text && prevParent?.text) {
      merged.parent = prevParent;
      merged.is_reply = true;
    }
    existing[id] = merged;
  }
  const ordered = Object.values(existing).sort((a, b) =>
    String(b.published_at || b.scraped_at || "").localeCompare(String(a.published_at || a.scraped_at || ""))
  );
  return [ordered, newCount];
}

export function savePosts(root: string, posts: Post[]): void {
  const { postsJsonl, postsCsv, data } = paths(root);
  mkdirSync(data, { recursive: true });
  writeFileSync(
    postsJsonl,
    posts.map((post) => JSON.stringify(post)).join("\n") + (posts.length ? "\n" : ""),
    "utf8"
  );
  const fields = [
    "id",
    "code",
    "url",
    "username",
    "text",
    "published_at",
    "like_count",
    "reply_count",
    "repost_count",
    "query",
    "source",
    "scraped_at",
  ];
  const rows = [
    fields.join(","),
    ...posts.map((post) =>
      fields
        .map((field) => `"${String(post[field] ?? "").replace(/"/g, '""')}"`)
        .join(",")
    ),
  ];
  writeFileSync(postsCsv, rows.join("\n"), "utf8");
}

export function saveRun(root: string, summary: Record<string, unknown>): string {
  const stamp = new Date().toISOString().replace(/[-:]/g, "").replace(/\.\d{3}Z$/, "Z");
  const file = join(paths(root).runs, `${stamp}.json`);
  writeFileSync(file, JSON.stringify(summary, null, 2), "utf8");
  return file;
}

export function lookbackPosts(posts: Post[], hours: number): Post[] {
  const cutoff = Date.now() - Math.max(1, hours) * 3600 * 1000;
  return posts.filter((post) => {
    const raw = String(post.published_at || "");
    if (!raw) {
      return false;
    }
    const ms = Date.parse(raw.replace("Z", "+00:00"));
    return !Number.isNaN(ms) && ms >= cutoff;
  });
}
